src/imap/select_folder.rs

```diff
@@ -139,19 +140,21 @@ impl Imap {
         &mut self,
         context: &Context,
         folder: &str,
-    ) -> Result<NewlySelected> {
+    ) -> anyhow::Result<NewlySelected> {
         match self.select_folder(context, Some(folder)).await {
             Ok(newly_selected) => Ok(newly_selected),
             Err(err) => match err {
-                Error::NoFolder(_) => {
+                Error::NoFolder(..) => {
                     if let Some(ref mut session) = self.session {
-                        session.create(folder).await?;
+                        session.create(folder).await.with_context(|| {
+                            format!("Couldn't select folder ('{}'), then create() failed", err)
+                        })?;
                     } else {
-                        return Err(Error::NoSession);
+                        return Err(Error::NoSession.into());
```

Comment:
Since you are returning `anyhow` error anyway, you can use
```
let session = self.session.as_mut().context("No IMAP session")?;
```
to avoid `if let`.